src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -269,12 +276,7 @@ public static void AssertValidVariableValue(ISchema schema, IGraphType type, obj
                 var value = astArguments?.ValueFor(arg.Name);
                 var type = arg.ResolvedType;
 
-                var coercedValue = CoerceValue(schema, type, value, variables) ?? arg.DefaultValue;
-
-                if (coercedValue != null)
-                {
-                    values[arg.Name] = coercedValue;
-                }
+                values[arg.Name] = CoerceValue(schema, type, value, variables, arg.DefaultValue);
```

Comment:
Previously, only arguments that were assigned a value were stored in the dictionary.  Now all arguments are stored in the dictionary after being assigned the correct default value.  This could represent a semantic difference, if someone is attempting to use `HasArgument` to determine if an argument was supplied or not.

However, not storing null values is semantically wrong as well, since supplying a value of `null` should result in `HasArgument` returning `true`.

And finally, not storing values that are not supplied doesn't work either, because it breaks the issue raised in #1127 where `context.Arguments` did not contain entries for arguments that should be set to their default values.

Ideally, I think that all fields should be stored in `context.Arguments`, but `HasArgument` should somehow be able to return `false` if the value was not supplied to the field (whether a default value was supplied or not) or, when it references a variable, not specified by either the inputs or a variable's default value.  This would be a bit more complicated to implement.

Reply:
> However, not storing null values is semantically wrong as well, since supplying a value of `null` should result in `HasArgument` returning `true`.

Agree.

> Ideally, I think that all fields should be stored in context.Arguments, but HasArgument should somehow be able to return false if the value was not supplied to the field (whether a default value was supplied or not) or, when it references a variable, not specified by either the inputs or a variable's default value. This would be a bit more complicated to implement.

`Dictionary<string, object> GetArgumentValues` 
->
`Dictionary<string, ArgumentValue> GetArgumentValues`

```c#
public struct ArgumentValue
{
  object Value { get; }
  ArgumentSource Source { get; } (or ArgumentKind)
}

public enum ArgumentSource (or ArgumentKind)
{
  Argument, (or Literal, Explicit)
  FieldDefault,
  Variable
}
```

` public static bool HasArgument(this IResolveFieldContext context, string name)`
->
` public static bool HasArgument(this IResolveFieldContext context, string name, ArgumentSource source)`

\+
` public static bool HasArgument(this IResolveFieldContext context, string name)` **may** be just an overload for returning any argument kind.

Reply:
This looks great!  For 4.0 I assume?  I can incorporate these changes.

Reply:
I mean, we COULD put these changes behind internal members so HasArgument works properly for 3.3 ... but I was think I'd just implement it in 4.0

Reply:
Let's keep it for v4.